src/main/java/sg/ncl/MainController.java

```diff
@@ -3494,6 +3505,539 @@ public String adminEnergy(Model model,
+            message.append(TAG_ERRORS);
+            message.append(TAG_UL);
+            for (ObjectError objectError : result.getAllErrors()) {
+                FieldError fieldError = (FieldError) objectError;
+                message.append(TAG_LI);
+                message.append(fieldError.getField());
+                message.append(TAG_SPACE);
+                message.append(fieldError.getDefaultMessage());
+                message.append(TAG_LI_CLOSE);
+            }
+            message.append(TAG_UL_CLOSE);
+            attributes.addFlashAttribute(MESSAGE, message.toString());
+        } else {
+            DateTimeFormatter formatter = new DateTimeFormatterBuilder().appendPattern("MMM-yyyy").toFormatter();
+            YearMonth m_s = YearMonth.parse(query.getStart(), formatter);
+            YearMonth m_e = YearMonth.parse(query.getEnd(), formatter);
+            YearMonth m_e_m2 = m_e.minusMonths(2);
+            YearMonth m_active = m_e_m2.isBefore(m_s) ? m_e_m2 : m_s;
+
+            YearMonth counter = m_s;
+            while (!counter.isAfter(m_e)) {
+                String monthYear = counter.format(formatter);
+                utilizationMap.put(monthYear, new MonthlyUtilization(monthYear));
+                months.add(monthYear);
+                counter = counter.plusMonths(1);
+            }
+            List<ProjectDetails> projectsList = getProjects();
+
+            for (ProjectDetails project : projectsList) {
+                // compute active and inactive projects
+                differentiateProjects(newProjects, activeProjects, inactiveProjects, m_s, m_e, m_e_m2, m_active, project);
+
+                // monthly utilisation
+                computeMonthlyUtilisation(utilizationMap, formatter, m_s, m_e, project);
+
+                // usage statistics by category
+                totalCategoryUsage += getCategoryUsage(statsCategoryMap, m_s, m_e, project);
+
+                // usage statistics by academic institutes
+                totalAcademicUsage += getAcademicUsage(statsAcademicMap, m_s, m_e, project);
+            }
+        }
+
+        attributes.addFlashAttribute(KEY_QUERY, query);
+        attributes.addFlashAttribute("newProjects", newProjects);
+        attributes.addFlashAttribute("activeProjects", activeProjects);
+        attributes.addFlashAttribute("inactiveProjects", inactiveProjects);
+        attributes.addFlashAttribute("months", months);
+        attributes.addFlashAttribute("utilization", utilizationMap);
+        attributes.addFlashAttribute("statsCategory", statsCategoryMap);
+        attributes.addFlashAttribute("totalCategoryUsage", totalCategoryUsage);
+        attributes.addFlashAttribute("statsAcademic", statsAcademicMap);
+        attributes.addFlashAttribute("totalAcademicUsage", totalAcademicUsage);
+
+        return "redirect:/admin/statistics";
+    }
+
+    private void differentiateProjects(List<ProjectDetails> newProjects,
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Method has 8 parameters, which is greater than 7 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00107)